lib/hash_ring/ets.ex

```diff
@@ -41,202 +59,220 @@ defmodule ExHashRing.HashRing.ETS do
-          _ -> find_node_inner(table, gen, num_nodes, key)
-        end
+  @doc """
+  Retrieves the current set of nodes as tuples of {name, replicas} from the ring.
+  """
+  @spec get_nodes_with_replicas(atom) :: {:ok, [{binary, integer}]}
+  def get_nodes_with_replicas(name) do
+    GenServer.call(name, :get_nodes_with_replicas)
+  end
 
-      {:error, error} ->
-        {:error, error}
+  @doc """
+  Retrieves the current set of overrides from the ring.
+  """
+  @spec get_overrides(atom) :: {:ok, override_map}
+  def get_overrides(name) do
+    GenServer.call(name, :get_overrides)
+  end
 
-      _ ->
-        {:error, :invalid_ring}
-    end
+  @doc """
+  Stops the GenServer holding the HashRing.
+  """
+  @spec stop(name :: atom()) :: :ok
+  def stop(name) do
+    GenServer.stop(name)
   end
 
-  defp find_node_inner(table, gen, num_nodes, key) do
-    hash = Utils.hash(key)
 
-    case find_next_highest_item(table, gen, num_nodes, hash) do
-      {_, node} -> {:ok, node}
-      _ -> {:error, :invalid_ring}
-    end
+  @doc """
+  Removes a node from the ring by its name.
+  """
+  @spec remove_node(atom, binary) :: {:ok, [{binary, integer}]} | {:error, :node_not_exists}
+  def remove_node(name, node_name) do
+    GenServer.call(name, {:remove_node, node_name})
   end
 
-  @spec find_nodes(atom, binary | integer, integer) :: {:ok, [binary]} | {:error, atom}
-  def find_nodes(name, key, num) do
-    hash = Utils.hash(key)
+  @doc """
+  Atomically remove multiple nodes from the ring by name
+  """
+  @spec remove_nodes(atom, [binary]) :: {:ok, [{binary, integer}]} | {:error, :node_not_exists}
+  def remove_nodes(name, node_names) do
+    GenServer.call(name, {:remove_nodes, node_names})
+  end
 
-    case Config.get(name) do
-      {:ok, {table, gen, num_nodes}} when num_nodes > 0 ->
-        nodes = do_find_nodes(table, gen, num_nodes, num, hash, [], 0)
+  @doc """
+  Replaces the nodes in the ring with a new set of nodes.
+  """
+  @spec set_nodes(atom, [binary | {binary, integer}]) :: {:ok, [{binary, integer}]}
+  def set_nodes(name, node_names) do
+    GenServer.call(name, {:set_nodes, node_names})
+  end
 
-        {:ok, nodes}
+  @doc """
+  Replaces the overrides in the ring with a new override map.
+  """
+  @spec set_overrides(atom, override_map) :: {:ok, override_map}
+  def set_overrides(name, overrides) do
+    GenServer.call(name, {:set_overrides, overrides})
+  end
 
-      {:ok, {table, gen, num_nodes, overrides}} when num_nodes > 0 and num > 0 ->
-        {found, found_length} =
-          case overrides do
-            %{^key => overrides} ->
-              {nodes, length} = Utils.take_max(overrides, num)
-              {Enum.reverse(nodes), length}
+  @doc """
+  Get the current ring generation
+  """
+  @spec get_ring_gen(atom) :: {:ok, integer} | :error
+  def get_ring_gen(name) do
+    with {:ok, {_current_table, _previous_table, ring_gen, _overrides}} <- Config.get(name) do
```

Comment:
can't you get this from `state.ring_gen`? If there's a race condition we should do some corrections.

Reply:
This is a client function so state is not available.